plugins/wacom-usb/fu-wac-module-bluetooth-id9.c

```diff
@@ -0,0 +1,264 @@
+					FuChunk *chunk,
+					FuProgress *progress,
+					GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GBytes) blob_chunk = NULL;
+
+	fu_byte_array_append_uint8(buf, phase);
+	g_byte_array_append(buf, fu_chunk_get_data(chunk), fu_chunk_get_data_sz(chunk));
+
+	blob_chunk = g_bytes_new(buf->data, buf->len);
+	return fu_wac_module_set_feature(self,
+					 FU_WAC_MODULE_COMMAND_DATA,
+					 blob_chunk,
+					 fu_progress_get_child(progress),
+					 FU_WAC_MODULE_WRITE_TIMEOUT,
+					 error);
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_blocks(FuWacModule *self,
+					 guint8 phase,
+					 GBytes *data,
+					 gsize block_len,
+					 FuProgress *progress,
+					 GError **error)
+{
+	g_autoptr(GPtrArray) chunks = fu_chunk_array_new_from_bytes(data, 0, 0, block_len);
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (gsize i = 0; i < chunks->len; i++) {
+		if (!fu_wac_module_bluetooth_id9_write_block(self,
+							     phase,
+							     g_ptr_array_index(chunks, i),
+							     progress,
+							     error))
+			return FALSE;
+
+		fu_progress_step_done(progress);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_extract_firmware(FuFirmware *firmware,
+					     GBytes **loader,
+					     GBytes **payload,
+					     GError **error)
+{
+	g_autoptr(GBytes) fw = fu_firmware_get_bytes(firmware, error);
+	const guint8 *blob;
+	gsize blob_len = 0;
+	gsize loader_len = 0;
+	gsize payload_len = 0;
+
+	if (fw == NULL)
+		return FALSE;
+
+	/* The firmware file is formatted as a 2 byte "length" field
+	 * followed by <length> bytes of loader code. The remainder
+	 * payload is firmware data.
+	 */
+	blob = g_bytes_get_data(fw, &blob_len);
+	if (blob_len < 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid firmware size");
+		return FALSE;
+	}
+
+	loader_len = fu_memread_uint16(blob, G_BIG_ENDIAN);
+	if (loader_len > blob_len - 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid firmware loader size");
+		return FALSE;
+	}
+	*loader = fu_bytes_new_offset(fw, 2, loader_len, NULL);
+
+	payload_len = blob_len - 2 - loader_len;
+	*payload = fu_bytes_new_offset(fw, 2 + loader_len, payload_len, NULL);
+
+	return TRUE;
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_firmware(FuDevice *device,
+					   FuFirmware *firmware,
+					   FuProgress *progress,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuWacModule *self = FU_WAC_MODULE(device);
+	const guint8 buf_start[] = {FU_WAC_MODULE_BLUETOOTH_ID9_START_NORMAL};
+	g_autoptr(GBytes) blob_start = g_bytes_new_static(buf_start, sizeof(buf_start));
+	g_autoptr(GBytes) loader = NULL;
+	g_autoptr(GBytes) payload = NULL;
+	g_autoptr(FuChunk) cmd = NULL;
+
+	if (!fu_wac_module_bluetooth_id9_extract_firmware(firmware, &loader, &payload, error))
+		return FALSE;
```

Comment:
i think rather than do this here you want to parse this in prepare_firmware() and then add the loader as a chiold image to the main firmware, and the payload as another child. Then you can just do `payload = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, error);` here. We want to do any parsing ahead of time whenever we can.

Reply:
Working on this. Looking at other plugins for examples of how to parse firmware into (child) images.

Reply:
ohh it's dead simple -- you can create a `g_autoptr(FuFirmware) img = fu_firmware_new()`, then `fu_firmware_set_id(img, FU_FIRMWARE_ID_PAYLOAD)` and then do  `fu_firmware_set_bytes(img, payload)` then a `fu_firmware_add_image(firmware, img)` -- then it shows up as a child firmware.

Reply:
Thanks! Updates have been made!